### ISSUE TITLE
DuplicateBasesError raised on kite-shaped inheritance graph with base class created by type()

### ISSUE THREAD

**Issue by BenediktSeidl** [PyCQA/astroid#633]
pylint crashes when linting a kite-shaped inheritance graph with base class created by `type()`. this seems to be a astroid problem and can be reproduced with following versions (empty fields were not tested):

| astroid version -> | 2.1.0 | 2.0.4 | 2.0.3 | 2.0 | master | 1.6.5 |
|--------|----|---|---|---|---|---|
| **python 3.7.2**    | | | | | crashes | crashes^1 |
| **python 3.7.0**    | crashes | crashes | crashes | crashes | crashes | crashes^1 |
| **python 3.6.8**    |  crashes   |   |    | crashes | crashes | crashes |
| **python 3.5.6**    | | | | | crashes | crashes |

<details><summary>1: exits with a different exception (click to expand)</summary>

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/astroid/decorators.py", line 89, in wrapped
    res = next(generator)
StopIteration

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/astroid/decorators.py", line 104, in wrapped
    raise StopIteration
StopIteration

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "<stdin>", line 18, in <module>
  File "/usr/local/lib/python3.7/site-packages/astroid/builder.py", line 282, in parse
    return builder.string_build(code, modname=module_name, path=path)
  File "/usr/local/lib/python3.7/site-packages/astroid/builder.py", line 159, in string_build
    return self._post_build(module, 'utf-8')
  File "/usr/local/lib/python3.7/site-packages/astroid/builder.py", line 173, in _post_build
    self.delayed_assattr(delayed)
  File "/usr/local/lib/python3.7/site-packages/astroid/builder.py", line 232, in delayed_assattr
    for inferred in node.expr.infer():
  File "/usr/local/lib/python3.7/site-packages/astroid/decorators.py", line 89, in wrapped
    res = next(generator)
  File "/usr/local/lib/python3.7/site-packages/astroid/bases.py", line 95, in _infer_stmts
    for inferred in stmt.infer(context=context):
  File "/usr/local/lib/python3.7/site-packages/astroid/context.py", line 71, in cache_generator
    for result in generator:
  File "/usr/local/lib/python3.7/site-packages/astroid/decorators.py", line 86, in wrapped
    generator = _func(node, context, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/astroid/inference.py", line 760, in infer_assign
    stmts = list(self.assigned_stmts(context=context))
  File "/usr/local/lib/python3.7/site-packages/astroid/protocols.py", line 304, in _arguments_infer_argname
    is_metaclass = isinstance(cls, nodes.ClassDef) and cls.type == 'metaclass'
  File "/usr/local/lib/python3.7/site-packages/astroid/scoped_nodes.py", line 1650, in _class_type
    if _is_metaclass(klass):
  File "/usr/local/lib/python3.7/site-packages/astroid/scoped_nodes.py", line 1636, in _is_metaclass
    if _is_metaclass(baseobj, seen):
  File "/usr/local/lib/python3.7/site-packages/astroid/scoped_nodes.py", line 1636, in _is_metaclass
    if _is_metaclass(baseobj, seen):
  File "/usr/local/lib/python3.7/site-packages/astroid/scoped_nodes.py", line 1619, in _is_metaclass
    for baseobj in base.infer():
RuntimeError: generator raised StopIteration
```

</details>

### Related Issues
* [pylint#2666](https://github.com/PyCQA/pylint/issues/2666) by @pdeblanc (this is where i took the minimal example and the headline)
* [#627 comment](https://github.com/PyCQA/astroid/issues/627#issuecomment-450108213) by @dzen

### Steps to reproduce

``` bash
# test.sh
pip install $1
python << __EOF__
import astroid
astroid.parse("""
A = type('A', (object,), {})

class B1(A):
    pass

class B2(A):
    pass

class C(B1, B2):
    pass

class D(C):
    def update(self):
        self.hello = 'hello'
""")
__EOF__
```

```
docker run -v$PWD:/workspace --rm python:3.5.6 bash /workspace/test.sh https://github.com/PyCQA/astroid/archive/master.zip
docker run -v$PWD:/workspace --rm python:3.7.0 bash /workspace/test.sh astroid==2.0.3
```

### Current behavior
crashes with 

```
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/astroid/decorators.py", line 31, in cached
    return cache[func]
KeyError: <bound method ClassDef.slots of <ClassDef.D l.13 at 0x7f75878789e8>>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "test.py", line 17, in <module>
    """)
  File "/usr/local/lib/python3.7/site-packages/astroid/builder.py", line 257, in parse
    return builder.string_build(code, modname=module_name, path=path)
  File "/usr/local/lib/python3.7/site-packages/astroid/builder.py", line 134, in string_build
    return self._post_build(module, 'utf-8')
  File "/usr/local/lib/python3.7/site-packages/astroid/builder.py", line 148, in _post_build
    self.delayed_assattr(delayed)
  File "/usr/local/lib/python3.7/site-packages/astroid/builder.py", line 214, in delayed_assattr
    if not _can_assign_attr(inferred, node.attrname):
  File "/usr/local/lib/python3.7/site-packages/astroid/builder.py", line 57, in _can_assign_attr
    slots = node.slots()
  File "/usr/local/lib/python3.7/site-packages/astroid/decorators.py", line 33, in cached
    cache[func] = result = func(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/astroid/scoped_nodes.py", line 2656, in slots
    slots = list(grouped_slots())
  File "/usr/local/lib/python3.7/site-packages/astroid/scoped_nodes.py", line 2642, in grouped_slots
    for cls in self.mro()[:-1]:
  File "/usr/local/lib/python3.7/site-packages/astroid/scoped_nodes.py", line 2735, in mro
    return self._compute_mro(context=context)
  File "/usr/local/lib/python3.7/site-packages/astroid/scoped_nodes.py", line 2716, in _compute_mro
    _verify_duplicates_mro(unmerged_mro, self, context)
  File "/usr/local/lib/python3.7/site-packages/astroid/scoped_nodes.py", line 94, in _verify_duplicates_mro
    mros=sequences, cls=cls, context=context)
astroid.exceptions.DuplicateBasesError: Duplicates found in MROs (D), (C, B1, A, B2, A, object), (C) for <ClassDef.D l.13 at 0x7f75878789e8>.
```

### Expected behavior
do not crash

*edit: version 1.6.5 crashes with all python versions*

**Comment by PCManticore** [comment]
@BenediktSeidl Thanks for creating this issue. I'm aware of the bug from `pylint`, unfortunately I didn't have time yet to investigate/fix.

**Comment by awarecan** [comment]
I am new to astroid code base. I think I pinned down the root cause after some debugging

```python
A = type('A', (object,), {})
```
`A` has been parsed as `Assign` not `ClassDef`, therefore `ClassDef(B1).bases[0] != ClassDef(B2).bases[0]` although their names are all `A`.

So that in the mro merger function two `A` be treated as different class and not be merged.

However I don't have idea how to resolve it.

**Comment by awarecan** [comment]
I successfully recreated the failure I saw in `aiohttp` library (>3.5.0)
A minimal example to trigger the issue
```python
# aiohttp_test.py
import aiohttp
class MyResource(aiohttp.web_urldispatcher.AbstractResource):
	pass
```
```
> pip install aiohttp==3.5.4 pylint=2.2.2
> pylint aiohttp_test.py
```

It caused by the typings wrapper around collection abc
https://github.com/python/cpython/blob/b8bcec35e01cac018f6ccfc8323d35886340efe0/Lib/typing.py#L1190

A more abstract example
```python
import astroid
astroid.parse("""
class Meta():
	def __init__(self, name):
		self.name = name

T = Meta('T')
S = Meta('S')

class C(T, S):
	def trigger_mro(self):
		self.name = 'C'
""")
```

**Comment by mthuurne** [comment]
I seem to be running into the same issue when checking the following code with PyLint:
```python
from typing import Generic, TypeVar

T = TypeVar('T')

class Base1(Generic[T]):
    pass

class Base2(Generic[T]):
    pass

class Multi(Base1, Base2):
    pass

class Concrete(Multi):
    def get_value(self):
        return self.value
```

It raises `DuplicateBasesError` on Python 3.7.2, but not on 3.6.5 or 3.5.6. I'm using astroid 2.1.0.

**Comment by pdeblanc** [comment]
Ping. My PR https://github.com/PyCQA/astroid/pull/634 fixes this issue. Can I please get a review?

**Comment by PCManticore** [comment]
Fixed with #634
